File: Assets/Scripts/P4.cs
using System.Diagnostics;
using System.IO;
using UnityEngine;
using System.Text;

public static class P4
{
    public static string P4Port { get; set; }
    public static string P4User { get; set; }

    public static bool IsMockMode = false;
    public static string MockLogPath;

    public static void SetMockMode(bool isMock)
    {
        IsMockMode = isMock;
        LogMockEvent($"Mock Mode {(isMock ? "Enabled" : "Disabled")}");
    }

    public static void LogMockEvent(string message)
    {
        if (string.IsNullOrEmpty(MockLogPath))
        {
            MockLogPath = Path.Combine(Application.persistentDataPath, "p4_mock_log.txt");
        }
        try { File.AppendAllText(MockLogPath, $"\n{message}\n"); } catch {}
    }

    public static (string Output, string Error) RunCommand(string arguments, string input = null)
    {
        if (string.IsNullOrEmpty(MockLogPath))
            MockLogPath = Path.Combine(Application.persistentDataPath, "p4_mock_log.txt");

        string logEntry = $"Command: p4 {arguments}\n";
        if (input != null) logEntry += $"Input Data:\n{input}\n";
        logEntry += "--------------------\n";

        try { File.AppendAllText(MockLogPath, logEntry); }
        catch (System.Exception e) { UnityEngine.Debug.LogError("Log Write Error: " + e.Message); }

        if (IsMockMode)
        {
            bool isReadOnly = arguments.Contains("group -o") || arguments.Contains("users") || arguments.Contains("login");
            
            if (!isReadOnly)
            {
                return ("Mock Command Success (Logged Only)", "");
            }
        }

        var startInfo = new ProcessStartInfo
        {
            FileName = "p4",
            Arguments = arguments,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            RedirectStandardInput = (input != null),
            UseShellExecute = false,
            CreateNoWindow = true,
        };

        if (!string.IsNullOrEmpty(P4Port)) startInfo.EnvironmentVariables["P4PORT"] = P4Port;
        if (!string.IsNullOrEmpty(P4User)) startInfo.EnvironmentVariables["P4USER"] = P4User;

        using (var process = Process.Start(startInfo))
        {
            if (input != null)
            {
                process.StandardInput.Write(input);
                process.StandardInput.Close();
            }

            string output = process.StandardOutput.ReadToEnd();
            string error = process.StandardError.ReadToEnd();
            process.WaitForExit();

            if (!string.IsNullOrEmpty(output)) UnityEngine.Debug.Log($"P4 Output: {output}");
            if (!string.IsNullOrEmpty(error)) UnityEngine.Debug.LogError($"P4 Error: {error}");

            return (output, error);
        }
    }
}

File: backend/p4.py
import subprocess
import os
import datetime

class P4Wrapper:
    P4PORT = ""
    P4USER = ""
    
    # Mock Settings
    MOCK_MODE = True
    MOCK_LOG_PATH = "p4_mock_log.txt"

    @staticmethod
    def LogMockEvent(message):
        """Allows external scripts to write headlines to the mock log."""
        try:
            with open(P4Wrapper.MOCK_LOG_PATH, "a") as f:
                f.write(f"\n{message}\n")
        except: pass

    @staticmethod
    def ExecuteRealCommand(arguments, input_data=None):
        """Helper to run the actual subprocess command."""
        env = os.environ.copy()
        if P4Wrapper.P4PORT: env["P4PORT"] = P4Wrapper.P4PORT
        if P4Wrapper.P4USER: env["P4USER"] = P4Wrapper.P4USER

        startupinfo = None
        if os.name == 'nt':
            startupinfo = subprocess.STARTUPINFO()
            startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW

        try:
            cmd = ["p4"] + arguments.split()
            process = subprocess.Popen(
                cmd,
                stdin=subprocess.PIPE if input_data else None,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                env=env,
                shell=False, 
                startupinfo=startupinfo
            )
            
            stdout, stderr = process.communicate(input=input_data)
            return stdout, stderr
        except FileNotFoundError:
             return "", "Error: 'p4' command not found."
        except Exception as e:
            return "", str(e)

    @staticmethod
    def RunCommand(arguments, input_data=None):
        log_entry = f"Command: p4 {arguments}\n"
        if input_data:
            log_entry += f"Input Data:\n{input_data}\n"
        log_entry += "-" * 20 + "\n"

        try:
            with open(P4Wrapper.MOCK_LOG_PATH, "a") as f:
                f.write(log_entry)
        except: 
            pass

        if P4Wrapper.MOCK_MODE:
            if "group -o" in arguments or "users" in arguments or "login" in arguments:
                return P4Wrapper.ExecuteRealCommand(arguments, input_data)
            
            return "Mock Command Success (Logged Only)", ""
        
        return P4Wrapper.ExecuteRealCommand(arguments, input_data)
    
    @staticmethod
    def AddUserToGroup(username, group_name):
        # Fetch current group spec
        out, err = P4Wrapper.RunCommand(f"group -o {group_name}")
        if err and "group not found" not in err.lower(): 
            return False, f"Fetch Error: {err}"

        # Check if user is already in it
        if f"\t{username}" in out or f" {username}" in out:
            return True, "User already in group."

        # Modify the spec text
        if "Users:" in out:
            lines = out.splitlines()
            new_lines = []
            in_users = False
            added = False
            for line in lines:
                new_lines.append(line)
                if line.startswith("Users:"):
                    in_users = True
                elif in_users and line.strip() == "":
                    new_lines.insert(-1, f"\t{username}")
                    added = True
                    in_users = False
            
            if not added:
                new_lines.append(f"\t{username}")
            
            new_spec = "\n".join(new_lines)
        else:
            new_spec = out + f"\nUsers:\n\t{username}\n"

        # Save the modified spec
        o, e = P4Wrapper.RunCommand("group -i", input_data=new_spec)
        if e: return False, f"Update Error: {e}"
        return True, f"User {username} added to {group_name}"
    
    @staticmethod
    def SetPassword(username, password):
        # Correct syntax is 'p4 passwd username' (no -u flag)
        return P4Wrapper.RunCommand(f"passwd {username}", input_data=f"{password}\n{password}\n")